Port forward with multiple ports

bnookala opened the pull request:
Fixes #185 

Allows users to enter multiple port bindings:

ex: 8888:80 8889:81 or 80 81 (for random assigned ports)

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -35,15 +35,15 @@ function isFindResultFromDocument(obj: PortForwardFindPodsResult): obj is PodFro
 
 /**
- * Implements port-forwarding to a target pod in the `default` namespace.
+ * Implements port-forwarding to a target pod in the `default` namespace
```
fixed!

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -88,23 +88,23 @@ export async function portForwardKubernetes (explorerNode?: any): Promise<void>
-            placeHolder: "ex: 8888:5000, for a specific port mapping or 5000 to pick a random local port",
-            prompt: `Enter the colon (':') seperated local port to remote port mapping you'd like to port forward.`,
+            placeHolder: "ex: 8888:5000 8889:5001",
+            prompt: `Enter the colon seperated (':') port mappings. Format is: LOCAL:REMOTE`,
```
fixed!

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -88,23 +88,23 @@ export async function portForwardKubernetes (explorerNode?: any): Promise<void>
     try {
         portString = await host.showInputBox(<QuickPickOptions>{
-            placeHolder: "ex: 8888:5000, for a specific port mapping or 5000 to pick a random local port",
-            prompt: `Enter the colon (':') seperated local port to remote port mapping you'd like to port forward.`,
```
fixed!

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -117,22 +117,34 @@ async function promptForPort (podSelection: string): Promise<PortMapping> {
 
     // User provided only the target port
-    if (!portMapping.includes(':') && Number(portMapping)) {
+    if (!portPair.includes(':') && Number(portPair)) {
```
yeah this stuck out to me as kind of ugly when i was taking another look at it. i'll implement your suggestion

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -151,21 +163,31 @@ function validatePortMapping (portMapping: string): string {
+    if (!portPair.includes(':')) {
         return <PortMapping> {
-            targetPort: Number(portString),
+            targetPort: Number(portPair),
```
fixed

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -203,18 +225,38 @@ async function findPortForwardablePods (): Promise<PortForwardFindPodsResult> {
- * @returns The locally bound port that was used.
+ * Invokes kubectl port-forward
+ * @param podName The pod name
+ * @param portMapping The PortMapping objects. Each object contains the requested local and target ports
```
@itowlson there are two places where it can get assigned:

1. After port prompt, after validation, if the user has provided both the local and target port
2. If the user has *not* provided the local port at the port prompt step, we pick one for them in the `portForwardToPod` function.

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -203,18 +225,38 @@ async function findPortForwardablePods (): Promise<PortForwardFindPodsResult> {
+    const usedPortMappings = await Promise.all(portMapping.map(buildUsablePortPair));
+
+    const portPairStrings = usedPortMappings.map((usedPortPair: PortMapping) => { return `${usedPortPair.localPort}:${usedPortPair.targetPort}`; });
+    const usedPorts = usedPortMappings.map((usedPortPair: PortMapping) => { return usedPortPair.localPort; });
```
seems like typescript can't infer return types for anonymous functions without the return expression; it thinks `localPorts` (changed the name as you suggested below) is a `void[]`. co-ercing the type (ie `const localPorts: number[]  = …` doesn't seem to work either.

itowlson replied:
That is... very weird.  I often use `map` with expression lambdas and get perfectly good type inference for both the argument type and return type.  Is type inference failing at the `Promise.all` stage perhaps?

bnookala replied:
@itowlson possibly - I changed the binding for the `Promise.all` to co-erce the return type into a `PortMapping[]` ie:

`const usedPortMappings: PortMapping[] = await Promise.all(portMapping.map(buildUsablePortPair));`

but that still doesn't fix it

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -203,18 +225,38 @@ async function findPortForwardablePods (): Promise<PortForwardFindPodsResult> {
+    const usedPortMappings = await Promise.all(portMapping.map(buildUsablePortPair));
+
+    const portPairStrings = usedPortMappings.map((usedPortPair: PortMapping) => { return `${usedPortPair.localPort}:${usedPortPair.targetPort}`; });
+    const usedPorts = usedPortMappings.map((usedPortPair: PortMapping) => { return usedPortPair.localPort; });
```
fixed

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -203,18 +225,38 @@ async function findPortForwardablePods (): Promise<PortForwardFindPodsResult> {
+    const usedPorts = usedPortMappings.map((usedPortPair: PortMapping) => { return usedPortPair.localPort; });
 
-    console.log(`port forwarding to pod ${podName} at port ${targetPort}`);
+    const usedNamespace = namespace === undefined ? 'default' : namespace;
```
fixed

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -203,18 +225,38 @@ async function findPortForwardablePods (): Promise<PortForwardFindPodsResult> {
-    console.log(`port forwarding to pod ${podName} at port ${targetPort}`);
+    const usedNamespace = namespace === undefined ? 'default' : namespace;
+
+    usedPortMappings.forEach((usedPortPair: PortMapping) => {
```
fixed

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -203,18 +225,38 @@ async function findPortForwardablePods (): Promise<PortForwardFindPodsResult> {
+    });
+
+    kubectl.invokeInNewTerminal(`port-forward ${podName} ${portPairStrings.join(' ')} -n ${usedNamespace}`, PORT_FORWARD_TERMINAL);
+    return usedPorts;
```
fixed

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -223,10 +265,8 @@ export async function portForwardToPod (podName: string, portMapping: PortMappin
+    return {
+        targetPort: targetPort,
+        localPort: usedPort
+    } as PortMapping;
```
fixed

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -203,18 +225,38 @@ async function findPortForwardablePods (): Promise<PortForwardFindPodsResult> {
+export async function portForwardToPod (podName: string, portMapping: PortMapping[], namespace?: string): Promise<number[]> {
+    const usedPortMappings = await Promise.all(portMapping.map(buildUsablePortPair));
+
+    const portPairStrings = usedPortMappings.map((usedPortPair: PortMapping) => { return `${usedPortPair.localPort}:${usedPortPair.targetPort}`; });
```
fixed

bnookala commented on src/components/kubectl/port-forward.ts:
```diff
@@ -151,21 +163,31 @@ function validatePortMapping (portMapping: string): string {
- * Builds a PortMapping object from the captured user input.
- * @param portString The port string provided by the user.
+ * Builds and returns multiple PortMapping objects
+ * @param portString A validated, user provided string containing the port mappings
```
i'll leave this out for now

bnookala commented:
@itowlson I've addressed your requested changes. Also noted that we can do a bit of cleanup with the kubectl business, so I've removed the creation in the dashboard component and port-forward component. we know get kubectl via dependency injection \o/

itowlson commented on src/components/kubectl/port-forward.ts:
```diff
@@ -28,24 +29,25 @@ interface PodFromDocument {
  * tree view.
  */
-export async function portForwardKubernetes (explorerNode?: any): Promise<void> {
+export async function portForwardKubernetes (kubectl: Kubectl, explorerNode?: any, ): Promise<void> {
```
The trailing comma in the parameter list looks odd to me - is this intentional?

bnookala replied:
definitely unintentional. will fix and merge.
